src/util/scoped.rs

```diff
@@ -0,0 +1,361 @@
+use std::cell::RefCell;
+use std::fmt;
+use std::mem;
+use std::rc::Rc;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::thread;
+use std::io;
+
+use util::atomic_option::AtomicOption;
+
+#[doc(hidden)]
+trait FnBox {
+    fn call_box(self: Box<Self>);
+}
+
+impl<F: FnOnce()> FnBox for F {
+    fn call_box(self: Box<Self>) {
+        (*self)()
+    }
+}
+
+/// Like `std::thread::spawn`, but without the closure bounds.
+pub unsafe fn spawn_unsafe<'a, F>(f: F) -> thread::JoinHandle<()>
+where
+    F: FnOnce() + Send + 'a,
+{
+    let builder = thread::Builder::new();
+    builder_spawn_unsafe(builder, f).unwrap()
+}
+
+/// Like `std::thread::Builder::spawn`, but without the closure bounds.
+pub unsafe fn builder_spawn_unsafe<'a, F>(
+    builder: thread::Builder,
+    f: F,
+) -> io::Result<thread::JoinHandle<()>>
+where
+    F: FnOnce() + Send + 'a,
+{
+    use std::mem;
+
+    let closure: Box<FnBox + 'a> = Box::new(f);
+    let closure: Box<FnBox + Send> = mem::transmute(closure);
+    builder.spawn(move || closure.call_box())
+}
+
+
+pub struct Scope<'a> {
+    dtors: RefCell<Option<DtorChain<'a>>>,
+}
+
+struct DtorChain<'a> {
+    dtor: Box<FnBox + 'a>,
+    next: Option<Box<DtorChain<'a>>>,
+}
+
+enum JoinState {
+    Running(thread::JoinHandle<()>),
+    Joined,
+}
+
+impl JoinState {
+    fn join(&mut self) {
+        let mut state = JoinState::Joined;
+        mem::swap(self, &mut state);
+        if let JoinState::Running(handle) = state {
+            let res = handle.join();
+
+            if !thread::panicking() {
+                res.unwrap();
+            }
+        }
+    }
+}
+
+/// A handle to a scoped thread
+pub struct ScopedJoinHandle<T> {
+    inner: Rc<RefCell<JoinState>>,
+    packet: Arc<AtomicOption<T>>,
+    thread: thread::Thread,
+}
+
+/// Create a new `scope`, for deferred destructors.
+///
+/// Scopes, in particular, support [*scoped thread spawning*](struct.Scope.html#method.spawn).
+///
+/// # Examples
+///
+/// Creating and using a scope:
+///
+/// ```
+/// crossbeam_epoch::util::scoped::scope(|scope| {
+///     scope.defer(|| println!("Exiting scope"));
+///     scope.spawn(|| println!("Running child thread in scope"))
+/// });
+/// // Prints messages in the reverse order written
+/// ```
+pub fn scope<'a, F, R>(f: F) -> R
```

Comment:
I hope that adding this here can be a temporary solution until scoped threads are made into a proper crate or readded to the standard library.

Reply:
Yeah, we should really move them into a separate crate, perhaps `crossbeam-scope` (or `crossbeam-scoped`, not sure).
Shall we create this repository and start moving scoped threads there?

Reply:
I am thinking of the crate's name.  `scoped.rs` depends on `atomic_option.rs`, so it will provide more functionalities than just scoped threads.  Maybe we can name it `crossbeam-util` or `crossbeam-base`.  What do you think?

Reply:
Well, it doesn't need to provide more functionality than just scoped threads: `crossbeam-scoped` can simply depend on `crossbeam-option`.

That said, it might make sense to have something like `crossbeam-utils` instead of a bunch of tiny crates.

Reply:
I prefer `crossbeam-utils`: I'm worried of too many crates.

I made `crossbeam-utils` in my account (https://github.com/jeehoonkang/crossbeam-utils), and made this PR to use it.  If you all agree with this, I will move it to `crossbeam-rs`, and register it to crates.io.